src/java/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoir.java

```diff
@@ -355,42 +347,50 @@ private void rescaleIfNeeded(long now)
             {
                 try
                 {
-                    rescale(now);
+                    decayingBuckets = rescale(now);
                 }
                 finally
                 {
-                    decayLandmark = now;
                     rescaling.set(false);
                 }
             }
         }
     }
 
-    private void rescale(long now)
+    /**
+     * rescale races with update;
+     * If rescale is called at the same time as an update, and an update falls into an already
+     * processed bucket this update will be lost. We accept that for the sake of update performance.
+     * OTOH, if the update falls into a bucket that is yet to be processed, it will be added
+     * with a weight that is consistent with the rescaleFactor used by rescale, so the new rescaled
+     * buckets will all have values consistent with the new decay landmark.
+     */
+    private DecayingBuckets rescale(long now)
     {
-        // despite striping its safe to rescale each bucket individually
-        final double rescaleFactor = forwardDecayWeight(now);
-        for (int i = 0; i < decayingBuckets.length(); i++)
+        final double rescaleFactor = decayingBuckets.forwardDecayWeight(now);
+        DecayingBuckets newBuckets = new DecayingBuckets(now);
```

Comment:
same comment as for rebase - do you think it makes sense to compute a plain long array and then create atomic long array from it?

Reply:
Rescaling happens every 30 minutes.
Doesn't the compiler optimize this out? I mean - since the `newBuckets` is a local variable, then  we don't need anything extra to have the same guarantees.

That said, faster rescale means smaller race condition window, so it doesn't hurt to actually measure that. I'll add a benchmark and we'll see :)

Reply:
an ad hoc benchmark indeed shows a significant difference.
no reason to pessimize prematurely, so I applied the comment